File: One_trend/code_michael/functions/functions_old/multiscale_statistic.c

/* The two main functions compute
(1) the values of the kernel averages hat(psi)_T(u,h) for each (u,h) in grid
(2) the value of the multiscale statistic hat(Psi).
The functions are described in more detail below. */


#include <math.h> 
#include <R.h>
#include <Rmath.h> 
#include <stdlib.h> 
#include <stdio.h>


/* Auxiliary functions */


double awert(double x){
   /* This function provides an absolute value of a double x */
   if(x < 0)
     return(-x);
   else
    return(x);
}

double epanc(double x){
   /* This function calculates the value of the Epanechnikov kernel at point x */
   if(x > 1 || x < -1)
     return(0.0);
   else
     return (3.0 / 4.0) * (1 - x * x);
}

double s_t_0(double u, double h, int T){
   /* This function computes the kernel constant S_T0 needed for the local linear weights 
   of the kernel average hat(psi)_T(u,h) */ 
   int t;
   double result;
   result = 0;
   for(t = 0; t < T; t++){
      result += epanc(((t + 1) / (float)T - u) / h);
   }
   return(result / (T * h));
}

double s_t_1(double u, double h, int T){
   /* This function computes the kernel constant S_T1 needed for the local linear weights 
   of the kernel average hat(psi)_T(u,h) */ 
   int t;
   double result;
   result = 0;
   for(t = 0; t < T; t++){
      result += epanc(((t + 1) / (float)T - u) / h) * (((t + 1) / (float)T - u) / h);
   }
   return(result / (T * h));
}


/* Main functions */


double psi_average(double data[], int T, double u, double h){

   /* This function computes the kernel average hat(psi)_T(u, h) 

      Inputs: 
      data    time series of observations
      T       sample size
      u       location, point in [0,1]
      h       bandwidth

      Output: 
      result  value of hat(psi)_T(u, h) 
   */

   int t;
   double x, weight, nom, denom, result;
   nom = 0;
   denom = 0;

   for(t = 0; t < T; t++){
      x = (((t + 1) / (float)T - u) / h);
      weight = epanc(x) * (s_t_0(u, h, T) * x - s_t_1(u, h, T));
      nom += weight * data[t];
      denom += weight * weight;
   }
   
   result = nom/sqrt(denom);

   return(result);
}


void multiscale_statistic(double *data, int *T, double *grid, double *correct, int *N, double *sigmahat, double *psi_values, double *test_values, double *max_value){

   /* Inputs: 
      data	   time series observations, vector of length T
      T        	   length of time series
      grid         grid vector (u_1,...,u_N, h_1,...,h_N)
      correct      vector of additive correction terms (lambda_1,...lambda_N)      
      N        	   length of the grid
      sigmahat	   estimator for sqrt of long-run error variance 

      Outputs: 
      psi_values   vector of individual test statistics (psi_average/sigmahat) for each (u,h) from the grid 
      test_values  vector of individual test statistics (|psi_average/sigmahat| - correct) for each (u,h) from the grid
      max_value    overall value of the multiscale statistic
 
   */

   int i;
   double vals;
 	
   for(i=0; i < N[0]; i++){
      vals = psi_average(data, T[0], grid[i], grid[i + N[0]]) / sigmahat[0];
      psi_values[i] = vals;
      test_values[i] = awert(vals) - correct[i];
      if(i == 0){
     	max_value[0] = test_values[0];
      } else if (test_values[i] > max_value[0]) {
    	max_value[0] = test_values[i];
      }
   }
}

File: One_trend/Not_needed/code_michael/functions/functions_old/multiscale_statistic.r
multiscale_statistic <- function(data, sigmahat, grid=NULL){

  # Function that calculates the multiscale statistic  
  #
  # Arguments:
  # data       vector of length T 
  # sigmahat   estimate of sqrt(long-run variance)
  # grid       list with the elements 
  #            grid[[1]]: grid of locations
  #            grid[[2]]: grid of bandwidths
  #            If grid=NULL, then a default grid is used.    
  #
  # Outputs: list with the elements  
  # grid.list  location-bandwidth grid under consideration 
  #            (with the same format as the argument 'grid')
  # psi_vals   vector of individual test statistics (hat(psi)/sigmahat) 
  #            for each point (u,h) in the location-bandwidth grid
  # test_vals  vector of individual test statistics (|hat(psi)/sigmahat| - correct) 
  #            for each point (u,h) in the location-bandwidth grid
  # max_val    overall value of the multiscale statistic

  T                  <- as.integer(length(data)) 
  sigmahat           <- as.double(sigmahat)
  storage.mode(data) <- "double"

  if(is.null(grid)) {
    GRID <- default_grid(T)
    grid.list <- list()
    grid.list[[1]] <- unique(GRID[,1])
    grid.list[[2]] <- unique(GRID[,2])    
  } else {
    GRID <- expand.grid(u = grid[[1]], h = grid[[2]])  
    grid.list <- grid
  }

  N                     <- as.integer(dim(GRID)[1]) 
  N.u                   <- as.integer(length(grid.list[[1]]))
  N.h                   <- as.integer(length(grid.list[[2]]))
  correct               <- sqrt(2*log(1/(2*GRID[,2])))
  GRID                  <- as.matrix(GRID)
  GRID                  <- as.vector(GRID) 
  storage.mode(GRID)    <- "double"
  storage.mode(correct) <- "double"
   
  psi_values  <- vector(mode = "double", length = N)
  test_values <- vector(mode = "double", length = N)
  max_value   <- vector(mode = "double", length = 1)
  
  result <- .C("multiscale_statistic", data, T, GRID, correct, N, sigmahat, psi_values, test_values, max_value)

  return(list(grid=grid.list,psi_vals=matrix(result[[7]],nrow=N.h,byrow=TRUE),test_vals=matrix(result[[8]],nrow=N.h,byrow=TRUE),max_val=result[[9]]))
}

